File: docker/golang/02-memory-bandwidth/main.go
package main

import (
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"log/slog"
	"net/http"
	"os"
	"strconv"
	"time"
)

// ── Response ─────────────────────────────────────────────────────────────────
type MemBwResponse struct {
	Runtime   string `json:"runtime"`
	SizeKb    int    `json:"size_kb"`
	Sha256    string `json:"sha256"`
	ElapsedUs int64  `json:"elapsed_us"`
}

// ── Handler ───────────────────────────────────────────────────────────────────
func memBwHandler(w http.ResponseWriter, r *http.Request) {
	q := r.URL.Query()

	sizeKb := 64
	if raw := q.Get("size_kb"); raw != "" {
		if v, err := strconv.Atoi(raw); err == nil && v > 0 {
			if v > 10240 {
				v = 10240
			}
			sizeKb = v
		}
	}
	noHash := q.Get("no_hash") == "true" || q.Get("no_hash") == "1"

	start := time.Now()

	// Allocate and fill buffer with deterministic pattern.
	buf := make([]byte, sizeKb*1024)
	for i := range buf {
		buf[i] = byte(i & 0xFF)
	}

	hashStr := ""
	if !noHash {
		sum := sha256.Sum256(buf)
		hashStr = hex.EncodeToString(sum[:])
	}

	elapsed := time.Since(start).Microseconds()

	slog.Info("membw completed", "size_kb", sizeKb, "elapsed_us", elapsed)

	resp := MemBwResponse{
		Runtime:   "golang-docker",
		SizeKb:    sizeKb,
		Sha256:    hashStr,
		ElapsedUs: elapsed,
	}

	w.Header().Set("Content-Type", "application/json")
	if err := json.NewEncoder(w).Encode(resp); err != nil {
		slog.Error("encode error", "err", err)
	}
}

func healthHandler(w http.ResponseWriter, _ *http.Request) {
	w.WriteHeader(http.StatusOK)
}

// ── Entry point ───────────────────────────────────────────────────────────────
func main() {
	logger := slog.New(slog.NewJSONHandler(os.Stdout, &slog.HandlerOptions{
		Level: slog.LevelInfo,
	}))
	slog.SetDefault(logger)

	mux := http.NewServeMux()
	mux.HandleFunc("GET /membw", memBwHandler)
	mux.HandleFunc("GET /health", healthHandler)

	addr := ":8080"
	slog.Info("server starting", "addr", addr)

	if err := http.ListenAndServe(addr, mux); err != nil {
		slog.Error("server error", "err", err)
		os.Exit(1)
	}
}

File: docker/rust/02-memory-bandwidth/src/main.rs
use axum::{
    extract::Query,
    http::StatusCode,
    response::Json,
    routing::get,
    Router,
};
use serde::{Deserialize, Serialize};
use sha2::{Digest, Sha256};
use std::time::Instant;
use tracing::info;

// ── Query params ─────────────────────────────────────────────────────────────
#[derive(Deserialize)]
struct MemBwParams {
    /// Buffer size in kilobytes. Default 64, max 10240.
    size_kb: Option<usize>,
    /// When non-zero, skips SHA-256 hashing (measures raw allocation cost).
    #[serde(default)]
    no_hash: Option<u8>,
}

// ── Response ─────────────────────────────────────────────────────────────────
#[derive(Serialize)]
struct MemBwResponse {
    runtime:    &'static str,
    size_kb:    usize,
    sha256:     String,
    elapsed_us: u128,
}

// ── Handler ───────────────────────────────────────────────────────────────────
async fn membw_handler(Query(params): Query<MemBwParams>) -> Json<MemBwResponse> {
    let size_kb = params.size_kb.unwrap_or(64).min(10240);
    let no_hash = params.no_hash.unwrap_or(0) != 0;
    let size_bytes = size_kb * 1024;

    let start = Instant::now();

    // Allocate and fill buffer with deterministic pattern.
    let buf: Vec<u8> = (0..size_bytes).map(|i| (i & 0xFF) as u8).collect();

    let sha256 = if no_hash {
        String::new()
    } else {
        let mut hasher = Sha256::new();
        hasher.update(&buf);
        hex::encode(hasher.finalize())
    };

    let elapsed_us = start.elapsed().as_micros();
    info!(size_kb, elapsed_us, "membw completed");

    Json(MemBwResponse {
        runtime: "rust-docker",
        size_kb,
        sha256,
        elapsed_us,
    })
}

async fn health_handler() -> StatusCode {
    StatusCode::OK
}

// ── Entry point ───────────────────────────────────────────────────────────────
#[tokio::main]
async fn main() {
    tracing_subscriber::fmt()
        .with_env_filter(
            tracing_subscriber::EnvFilter::from_default_env()
                .add_directive("memory_bandwidth=info".parse().unwrap()),
        )
        .json()
        .init();

    let app = Router::new()
        .route("/membw",     get(membw_handler))
        .route("/health", get(health_handler));

    let addr = "0.0.0.0:8080";
    info!("listening on {addr}");
    let listener = tokio::net::TcpListener::bind(addr).await.unwrap();
    axum::serve(listener, app).await.unwrap();
}
